src/mono/mono/metadata/class-init.c

```diff
@@ -4189,6 +4196,100 @@ mono_class_set_runtime_vtable (MonoClass *klass, MonoVTable *vtable)
 	klass->runtime_vtable = vtable;
 }
 
+static int
+index_of_class (MonoClass *needle, MonoClass **haystack, int haystack_size) {
+	for (int i = 0; i < haystack_size; i++)
+		if (haystack[i] == needle)
+			return i;
+
+	return -1;
+}
+
+static void
+concat_variance_search_table (MonoClass *klass, MonoClass **buf, int buf_size, int *buf_count, MonoClass *current) {
+	MonoClass **table;
+	int table_size;
+	mono_class_get_variance_search_table (current, &table, &table_size);
+	if (!table_size || !table)
+		return;
+
+	for (int i = 0; i < table_size; i++) {
+		MonoClass *iface = table[i];
+		if (index_of_class (iface, buf, *buf_count) >= 0)
+			continue;
+
+		g_assert (*buf_count < buf_size);
+		buf[*buf_count] = table[i];
+		(*buf_count) += 1;
+	}
+}
+
+static void
+build_variance_search_table_inner (MonoClass *klass, MonoClass **buf, int buf_size, int *buf_count, MonoClass *current) {
+	if (!m_class_is_interfaces_inited (current)) {
+		ERROR_DECL (error);
+		mono_class_setup_interfaces (current, error);
+		return_if_nok (error);
+	}
+	guint c = m_class_get_interface_count (current);
+	if (c) {
+		MonoClass **ifaces = m_class_get_interfaces (current);
+		for (guint i = 0; i < c; i++) {
+			MonoClass *iface = ifaces [i];
+			// Avoid adding duplicates.
+			if (index_of_class (iface, buf, *buf_count) >= 0)
+				continue;
+
+			if (mono_class_has_variant_generic_params (iface)) {
+				g_assert (*buf_count < buf_size);
+				buf[*buf_count] = iface;
+				(*buf_count) += 1;
+			}
+
+			concat_variance_search_table (klass, buf, buf_size, buf_count, iface);
+		}
+	}
+}
+
+// Only call this with the loader lock held
+static void
+build_variance_search_table (MonoClass *klass) {
+	// FIXME: Is there a way to deterministically compute the right capacity?
```

Comment:
I think `MonoClass:interface_offsets_count` (the # of elts in `MonoClass:interfaces_packed`) is the total flattened interface list for the class.   

That should be an upper bound on the number of table entries you could possibly see, right?  I'm not sure if it removes any duplicates.

I think it should be inited by the time you build the table.